Add RequestInit members to Request domintro

domfarolino opened the pull request:
This PR adds the RequestInit members to the Request domintro box. I wasn't sure if I should be linking to the enums for things like RequestInit.{mode, credentials, etc} to make it easy for people to see the actual valid values of objects or not, so if that is necessary I'll adjust this for sure.

There are also a couple misc wording fixes I made for the lower content I added in an earlier PR.

Makes progress on https://github.com/whatwg/fetch/issues/543


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/fetch/620.html" title="Last updated on Jan 13, 2018, 12:36 PM GMT (0e4a7be)">Preview</a> | <a href="https://whatpr.org/fetch/620/fd28675...0e4a7be.html" title="Last updated on Jan 13, 2018, 12:36 PM GMT (0e4a7be)">Diff</a>

annevk commented on fetch.bs:
```diff
@@ -4860,8 +4860,59 @@ initially a new {{AbortSignal}} object.
+ {{Request}} object.
+
+ The optional <var>init</var> argument allows for setting properties appearing in {{RequestInit}}
+ via object members of the same name. These are the object members that can be used:
```
I prefer explicit `<p>` usage.

annevk commented on fetch.bs:
```diff
@@ -4860,8 +4860,59 @@ initially a new {{AbortSignal}} object.
+  <dd>Set to a string to set <var>request</var>'s {{Request/method}}.
+
+  <dt>{{RequestInit/headers}}
+  <dd>Set to a {{Headers}} object to set <var>request</var>'s {{Request/headers}}.
```
This can also be an array of two-item arrays.

annevk commented on fetch.bs:
```diff
@@ -4860,8 +4860,59 @@ initially a new {{AbortSignal}} object.
+
+  <dt>{{RequestInit/referrer}}
+  <dd>Set to a string whose value is a same-origin URL or the empty string to set
+  <var>request</var>'s <a>referrer</a>.
```
You can also set it to `about:client` to use the referrer of the global object.

annevk commented on fetch.bs:
```diff
@@ -4860,8 +4860,59 @@ initially a new {{AbortSignal}} object.
+  <var>request</var>'s {{Request/cache}}.
+
+  <dt>{{RequestInit/redirect}}
+  <dd>Set to a string indicating whether or not <var>request</var> should automatically  follow
```
follow should not be preceded by two spaces.

domenic commented:
Haven't looked at the PR much, but enums could get their own domintro (via their associated concepts), like https://fetch.spec.whatwg.org/#request-service-workers-mode or especially https://fetch.spec.whatwg.org/#concept-request-cache-mode . Then you could link to them there.

domfarolino commented:
Ok I've rebased this PR and addressed Anne's comments! Sorry for the crazy delay on this by the way :(

I think the next step is to go ahead and create `<div class="note no-backref">`s for [mode](https://fetch.spec.whatwg.org/#concept-request-mode), [credentials](https://fetch.spec.whatwg.org/#concept-request-credentials-mode), and [redirect](https://fetch.spec.whatwg.org/#concept-request-redirect-mode) concepts (since as pointed out, **cache** already has a nice description), as Domenic mentioned. Then I'll update this PR to link from the domintro box to the associated concepts which have their new descriptions.

annevk commented:
Don't worry about the delay, I'm very grateful you're taking this on. Plan sounds good to me.

domfarolino commented:
Alright I've added some domintro boxes in the #concept-request section for the enums which should, barring any request changes, finish this PR off.

annevk commented on fetch.bs:
```diff
@@ -1001,11 +1001,34 @@ unset.
+ <dl>
+  <dt>"<code>same-origin</code>"
+  <dd>Used to ensure requests are made to same-origin URLs. <a for=/>Fetch</a> will throw an error
+  if the request is not made to a same-origin URL.
```
`throw an error` -> `return a <a>network error</a>`

annevk replied:
This seems like it should be changed throughout.

domfarolino replied:
Ah ok, because it is the fetch *method* that throws by rejecting promise, not the algorithm (which just returns network error which the method can later make use of)...

annevk commented on fetch.bs:
```diff
@@ -1001,11 +1001,34 @@ unset.
+  if the request is not made to a same-origin URL.
+
+  <dt>"<code>cors</code>"
+  <dd>Ensures that the request being made is a <a>CORS request</a>. Fetch will throw an error if
```
I would say "Makes the request a CORS request".

annevk commented on fetch.bs:
```diff
@@ -1001,11 +1001,34 @@ unset.
+  <dt>"<code>no-cors</code>"
+  <dd>Restricts requests to using <a>CORS-safelisted methods</a> and
+  <a>CORS-safelisted request-headers</a>. Upon success, fetch will return a
+  <a>opaque filtered response</a>.
```
an opaque*

annevk commented on fetch.bs:
```diff
@@ -1001,11 +1001,34 @@ unset.
+ <p>Even though the default <a for=/>request</a>
+ <a for=request>mode</a> is "<code>no-cors</code>", standards are highly
+ discouraged from using it for new features. It is rather unsafe. "<code>navigate</code>" and
+ "<code>websocket</code>" are special values for the HTML Standard.
```
Given that we now mention navigate and websocket above, I don't think we need to mention them here again.

annevk commented on fetch.bs:
```diff
@@ -1021,14 +1044,27 @@ one or more event listeners are registered on an {{XMLHttpRequestUpload}} object
+<div class="note no-backref">
+ <dl>
+  <dt>"<code>omit</code>"
+  <dd>Used to ensure credentials are not included with this request.
```
"Excludes credentials from the request". In the sense that it doesn't really ensure it, that's just what happens.

annevk commented on fetch.bs:
```diff
@@ -1091,6 +1127,20 @@ Unless stated otherwise, it is unset.
+  <dd>Used to follow all redirects incurred when fetching a resource.
+
+  <dt>"<code>error</code>"
+  <dd>This mode ensures that fetch throws an error when a request for a resource is redirected.
```
It might make more sense to talk about the response being a redirect. Since no redirect takes place. Same below.

annevk commented on fetch.bs:
```diff
@@ -4903,8 +4953,60 @@ initially a new {{AbortSignal}} object.
+
+  <dt>{{RequestInit/referrer}}
+  <dd>Set to a string whose value is a same-origin URL, "<code>about:client</code>", or the empty
+  string to set <var>request</var>'s <a>referrer</a>.
```
comma after empty string?

domenic commented on fetch.bs:
```diff
@@ -4903,8 +4951,60 @@ initially a new {{AbortSignal}} object.
+  <dd>Set to a string to set <var>request</var>'s {{Request/method}}.
+
+  <dt>{{RequestInit/headers}}
+  <dd>Set to a {{Headers}} object or an array of two-item arrays to set <var>request</var>'s
```
"a Headers object, an object literal, or an array of two-item arrays"

domfarolino replied:
Ah I didn't know it could be an object literal!

domenic commented on fetch.bs:
```diff
@@ -4903,8 +4951,60 @@ initially a new {{AbortSignal}} object.
+  {{Request/headers}}.
+
+  <dt>{{RequestInit/body}}
+  <dd>Set to a {{Body}} object or null to set <var>request</var>'s <a>body</a>.
```
Not quite correct. It's actually a BodyInit. Either link to BodyInit, or maybe spell out what a BodyInit is.

domfarolino replied:
I opted for linking if that's ok. I figured that soon `BodyInit` will have its own domintro so what a developer might get linked too shouldn't be too tough to understand.

domenic commented on fetch.bs:
```diff
@@ -4903,8 +4951,60 @@ initially a new {{AbortSignal}} object.
+
+ <dl>
+  <dt>{{RequestInit/method}}
+  <dd>Set to a string to set <var>request</var>'s {{Request/method}}.
```
Maybe omit "Set to" in all of these?

domenic commented on fetch.bs:
```diff
@@ -4903,8 +4951,60 @@ initially a new {{AbortSignal}} object.
+  {{Request/redirect}}.
+
+  <dt>{{RequestInit/integrity}}
+  <dd>Set to a cryptographic has of the resource to be fetched by <var>request</var>. Sets
```
s/has/hash

domenic commented on fetch.bs:
```diff
@@ -4903,8 +4951,60 @@ initially a new {{AbortSignal}} object.
+  <dd>Set to an {{AbortSignal}} to set <var>request</var>'s {{Request/signal}}.
+
+  <dt>{{RequestInit/window}}
+  <dd>Can only be null. Used to set <var>request</var>'s internal <a>window</a> object to
```
Not an object. I'd maybe say "Used to de-associate the request from any {{Window}}"?

domenic commented on fetch.bs:
```diff
@@ -1001,11 +1001,32 @@ unset.
+  <dd>This is a special mode used only when navigating between documents.
+
+  <dt>"<code>websocket</code>"
+  <dd>This is a special mode used only when establishing a WebSocket connection.
```
Maybe link this to the "establish a WebSocket connection" algorithm.

domenic commented on fetch.bs:
```diff
@@ -1001,11 +1001,32 @@ unset.
+  <a>opaque filtered response</a>.
+
+  <dt>"<code>navigate</code>"
+  <dd>This is a special mode used only when navigating between documents.
```
Maybe `<a lt="navigate">navigating</a>`, which should automatically link to https://html.spec.whatwg.org/#navigate

domenic commented on fetch.bs:
```diff
@@ -1091,6 +1125,20 @@ Unless stated otherwise, it is unset.
+<div class="note no-backref">
+ <dl>
+  <dt>"<code>follow</code>"
+  <dd>Used to follow all redirects incurred when fetching a resource.
```
s/Used to follow/Follow

domenic commented on fetch.bs:
```diff
@@ -1091,6 +1125,20 @@ Unless stated otherwise, it is unset.
+  <dd>Used to follow all redirects incurred when fetching a resource.
+
+  <dt>"<code>error</code>"
+  <dd>This mode makes fetch return a <a>network error</a> when a request is met with a redirect.
```
s/This mode makes fetch return/Return

domenic commented on fetch.bs:
```diff
@@ -1091,6 +1125,20 @@ Unless stated otherwise, it is unset.
+  <dd>This mode makes fetch return a <a>network error</a> when a request is met with a redirect.
+
+  <dt>"<code>manual</code>"
+  <dd>This mode is used to retrieve an <a>opaque-redirect filtered response</a> when a request is
```
s/This mode is used to retrieve/Retrieves

domenic commented on fetch.bs:
```diff
@@ -4903,8 +4951,59 @@ initially a new {{AbortSignal}} object.
+ {{Request}} object.
+
+ <p>The optional <var>init</var> argument allows for setting properties appearing in {{RequestInit}}
+ via object members of the same name. These are the object members that can be used:</p>
```
I think this could be tightened up to not reference the RequestInit dictionary; now that this documentation exists, developers shouldn't need to know about it. E.g. "The optional _init_ argument is an object whose properties can be set as follows:"

domfarolino replied:
That sounds good to me

domenic commented on fetch.bs:
```diff
@@ -4903,8 +4951,59 @@ initially a new {{AbortSignal}} object.
+  <dd>An {{AbortSignal}} to set <var>request</var>'s {{Request/signal}}.
+
+  <dt>{{RequestInit/window}}
+  <dd>Can only be null. Used to set disassociate <var>request</var> from any {{Window}}.
```
s/set disassociate/disassociate

annevk commented:
Thanks @domfarolino! (I found a couple more very minor nits I decided to fix myself.)
